README.md
![Lifecycle:Experimental](https://img.shields.io/badge/Lifecycle-Experimental-339999)

# Vecnet

`vecnet` provides tools to vectorize a road network either from a binary map or a probability map derived from machine learning or any other method.  The algorithm is expected to be robust to gaps (false negative) and false positive. This package is experimental and must be considered as a proof of concept. It contains the implementation of the algorithm described in:

> Roussel, J., Bourdon, J., Morley, I. D., Coops, N. C., & Achim, A. (2023). Vectorial and topologically valid segmentation of forestry road networks from ALS data. International Journal of Applied Earth Observation and Geoinformation, 118, 103267. https://doi.org/10.1016/j.jag.2023.103267

The version [0.1.0](https://github.com/Jean-Romain/vecnet/releases/tag/v0.1.0) corresponds to the state of the software when submitted to the journal. Next versions include enhancements and bug fixes.

## Installation

``` r
remotes::install_github("Jean-Romain/vecnet")
```


NEWS.md
# vecnet 0.1.2

* New parameter `smooth` in `vectorize_network()` to return a smoothed network with few nodes
* Fix a bugs that prevent connection between the road being vectorized and an already vectorized road.
* Change the dataset shipped with the package

# vecnet 0.1.1

* New parameters `th_conductivity` and `partial_gap_size`
* Improve driving in sharp turns: 
   * The trace left on the current road has a value 0.05 instead of 0
   * The orientation of the sight of view is computed from the last 50% of the previous section drive instead of the whole section. This way the next part of the road is more likely to be oriented close to 0 degree.
* Added a `NEWS.md` file to track changes to the package.

# vectnet 0.1.0

* Submitted to International Journal of Applied Earth Observation and Geoinformation


R/lcp.R
#' @importClassesFrom  gdistance TransitionLayer
transition <- function(conductivity, directions = 8, geocorrection = TRUE)
{
  x = conductivity
  use_terra = FALSE

  if (methods::is(x, "RasterLayer"))
  {
    bb = raster::extent(x)
    ncells = raster::ncell(x)
    val = raster::values(x)
  }
  else if (methods::is(x, "SpatRaster"))
  {
    bb = terra::ext(x)
    bb = raster::extent(bb[1:4])
    ncells = terra::ncell(x)
    val = terra::values(x)[,1]
    use_terra = TRUE
  }
  else
  {
    stop("Only RasterLayer and SpatRaster are supported")
  }

  symm = TRUE

  tr <- methods::new("TransitionLayer",
            nrows=as.integer(nrow(x)),
            ncols=as.integer(ncol(x)),
            extent=bb,
            crs=sp::CRS(),
            transitionMatrix = Matrix::Matrix(0, ncells,ncells),
            transitionCells = 1:ncells)

  transitionMatr <- gdistance::transitionMatrix(tr)
  Cells <- which(!is.na(val))

  if (use_terra)
    adj <- terra::adjacent(x, cells=Cells, pairs=TRUE, directions=directions)
  else
    adj <- raster::adjacent(x, cells=Cells, pairs=TRUE, directions=directions)

  if(symm)
    adj <- adj[adj[,1] < adj[,2],]

  dataVals <- cbind(val[adj[,1]], val[adj[,2]])
  transition.values <- rowMeans(dataVals)
  transition.values[is.na(transition.values)] <- 0

  if(!all(transition.values >= 0))
    warning("transition function gives negative values")

  transitionMatr[adj] <- as.vector(transition.values)
  if(symm)
    transitionMatr <- Matrix::forceSymmetric(transitionMatr)

  gdistance::transitionMatrix(tr) <- transitionMatr
  gdistance::matrixValues(tr) <- "conductance"


  #trans <- gdistance::transition(conductivity, transitionFunction = mean, directions = 8)
  trans = tr

  if (isTRUE(geocorrection))
    trans <- geocorrection(trans)

  return(trans)
}

# Modification, simplification and use terra of the function gdistance::geoCorrection()
geocorrection = function(x)
{
  adj <- gdistance::adjacencyFromTransition(x)
  cell1 = terra::xyFromCell(x, adj[,1])
  cell2 = terra::xyFromCell(x,adj[,2])

  distance = (cell1-cell2)^2
  distance = sqrt(distance[,1] + distance[,2])

  if(gdistance::matrixValues(x) == "conductance")
    correctionValues <- 1/distance
  else if(gdistance::matrixValues(x) == "resistance")
    correctionValues <- distance
  else
    stop("Internal error")

  i <- as.integer(adj[,1] - 1)
  j <- as.integer(adj[,2] - 1)
  xv <- as.vector(correctionValues) #check for Inf values!
  dims <- ncell(x)
  correctionMatrix <- new("dgTMatrix", i = i, j = j, x = xv, Dim = as.integer(c(dims,dims)))
  correctionMatrix <- (methods::as(correctionMatrix,"sparseMatrix"))

  if (is(gdistance::transitionMatrix(x), "dsCMatrix")) #isSymmetric?
    correctionMatrix <- Matrix::forceSymmetric(correctionMatrix)

  transitionCorrected <- correctionMatrix * gdistance::transitionMatrix(x)
  gdistance::transitionMatrix(x) <- transitionCorrected
  return(x)
}



R/smooth.R
smooth_network = function(net)
{
    smoothed = lapply(net, smooth_line)
    smoothed = do.call(c, smoothed)
    sf::st_crs(smoothed) = sf::st_crs(net)
    return(smoothed)
}

smooth_line = function(points)
{
  # Adjust a spline to create a smooth line from points
  xroad <- sf::st_coordinates(points)[,1]
  yroad <- sf::st_coordinates(points)[,2]

  if (length(xroad) < 4) return(sf::st_sfc(points))

  # Retrieve the "temporal" position along the line so we can order the points later
  xend <- xroad[length(xroad)]
  yend <- yroad[length(yroad)]
  xstart <- xroad[1]
  ystart <- yroad[1]
  dx  <- c(xroad, xend) - c(xstart, xroad)
  dy  <- c(yroad, yend) - c(ystart, yroad)
  dd  <- sqrt(dx*dx+dy*dy)
  dist <- cumsum(dd[-length(dd)])
  troad <- dist/dist[length(dist)]

  # Weigth the first and last node
  wroad <- rep(1, length(xroad))
  wroad[0] = 1e10
  wroad[length(wroad)] = 1e10

  ux <- stats::smooth.spline(troad, xroad, wroad, spar = 0.4, all.knots = TRUE)
  uy <- stats::smooth.spline(troad, yroad, wroad, spar = 0.4, all.knots = TRUE)

  # Sometime one point is missing in the spline for an unknown reason. If the output
  # does not have the same length than the input we resize the input to fit the output
  # and prevent failures
  rm <- FALSE
  if (length(ux$x) < length(troad))
  {
    rm <- troad %in% ux$x
    xroad = xroad[rm]
    yroad = yroad[rm]
    troad = troad[rm]
    wroad = wroad[rm]
  }

  # Maybe it is possible to have an extra point. This case is handled here but never observed
  if (length(ux$x) > length(troad))
  {
    print(dput(points))
    stop("Different length")
  }

  # The adjusted spline can be far from some outlier points. This is the role of the spline
  # to smooth the road. But if some important outliers are detected far from the spline (more than
  # 10 m) we remove those points and fit again the spline
  rm2 <- FALSE
  rmx <- abs(ux$y - xroad) < 10
  rmy <- abs(uy$y - yroad) < 10

  if (any(!rmx) | any(!rmy))
  {
    rm2   <- rmx & rmy
    xroad <- xroad[rm2]
    yroad <- yroad[rm2]
    troad <- troad[rm2]
    wroad <- wroad[rm2]

    if (length(xroad) > 3)
    {
      ux <- stats::smooth.spline(troad, xroad, w = wroad)
      uy <- stats::smooth.spline(troad, yroad, w = wroad)
    }
    else
    {
      rm2 <- FALSE
    }
  }

  if (!isFALSE(rm))  spline <- spline[rm,]
  if (!isFALSE(rm2)) spline <- spline[rm2,]

  spline <- cbind(ux$y, uy$y)
  spline[1,1] = xstart
  spline[1,2] = ystart
  spline[nrow(spline),1] = xend
  spline[nrow(spline),2] = yend
  spline <- sf::st_sfc(sf::st_linestring(spline))
  sf::st_crs(spline) <- st_crs(points)
  spline = sf::st_simplify(spline, dTolerance = 1)
  return(spline)
}


R/sf_utils.R
#' Get heading of both ends of a line
#'
#' Retrieve heading of both ends of a line, pointing away from it.
#'
#' @param line  line (\code{sfc} or \code{sfg} format)
#'
#' @return numeric of length 2 expressing respectively the head and tail heading in degrees
#' (range [-180°, 180°] using \code{atan2()}.
#' @noRd
st_ends_heading <- function(line)
{
  M <- sf::st_coordinates(line)
  i <- c(2, nrow(M) - 1)
  j <- c(1, -1)

  headings <- mapply(i, j, FUN = function(i, j) {
    Ax = M[i-j,1]
    Ay = M[i-j,2]
    Bx = M[i,1]
    By = M[i,2]
    atan2(Ay-By, Ax-Bx)*180/pi
  })
  names(headings) <- c("head", "tail")
  return(headings)
}


st_join_linestring = function(x,y)
{
  X = sf::st_coordinates(x)
  Y = sf::st_coordinates(y)
  M = rbind(X,Y)[,1:2]
  L = sf::st_sfc(sf::st_linestring(M))
  return(L)
}

#' Extend line by given distance
#'
#' Extend one or both ends of a line by a given distance.
#' No new vertices are added, instead, the first/last is moved
#' to its new location following the same heading as before.
#'
#' @param line  line (\code{sfc} or \code{sfg} format)
#' @param distance  numeric; distance by which the line will be extended. In case of \code{end = "BOTH"},
#' a vector of length 2 can be provided to extend respectively the head and tail end by different values.
#' @param end  character; (\code{BOTH}, \code{HEAD} or \code{TAIL}; define which end will be extended.
#'
#' @return Same line as input but now extended.
#' @noRd
st_extend_line <- function(line, distance, end = "BOTH")
{
  if (!(end %in% c("BOTH", "HEAD", "TAIL")) | length(end) != 1) stop("'end' must be 'BOTH', 'HEAD' or 'TAIL'")

  M <- sf::st_coordinates(line)[,-3]
  keep <- !(end == c("TAIL", "HEAD"))

  ends <- c(1, nrow(M))[keep]
  headings <- st_ends_heading(line)[keep] / 180 * pi
  distances <- if (length(distance) == 1) rep(distance, 2) else rev(distance[1:2])

  M[ends, 1:2] <- M[ends, 1:2] + distances[keep] * c(cos(headings), sin(headings))
  newline <- sf::st_linestring(M)

  # If input is sfc_LINESTRING and not sfg_LINESTRING
  if (is.list(line)) newline <- sf::st_sfc(newline, crs = sf::st_crs(line))

  return(newline)
}

#' @examples
#' f <- system.file("extdata", "sinuosity", package="vectnet")
#' m = sf::st_read(f)
#' m = sf::st_transform(m, 27572)
sinuosity <- function(x)
{
  lines <- x
  n <- length(lines)
  S <- numeric(n)
  for (i in 1:n)
  {
    line <- lines[i]

    if (sf::st_geometry_type(line) != "LINESTRING")
      stop(paste0("Geometry ", i, " is not a LINESTRING"))

    spline <- adjust_spline(line)
    S[i] = sf::st_length(line)/sf::st_length(spline)
  }

  round(S,2)
}

adjust_spline = function(points)
{
  # Adjust a spline to create a smooth line from points
  xroad <- sf::st_coordinates(points)[,1]
  yroad <- sf::st_coordinates(points)[,2]

  if (length(xroad) <= 4)
    return(points)

  n = length(xroad)
  troad <- cumsum(c(0,sqrt((xroad[-1]-xroad[-n])^2 + (yroad[-1]-yroad[-n])^2)))

  ux <- stats::smooth.spline(troad, xroad, spar = 0.5, all.knots = TRUE)
  uy <- stats::smooth.spline(troad, yroad, spar = 0.5, all.knots = TRUE)

  # Sometime one point is missing in the spline for an unknown reason. If the output
  # does not have the same length than the input we resize the input to fit the output
  # and prevent failures
  rm <- FALSE
  if (length(ux$x) < length(troad))
  {
    rm <- troad %in% ux$x
    xroad = xroad[rm]
    yroad = yroad[rm]
    troad = troad[rm]
  }

  # Maybe it is possible to have an extra point. This case is handled here but never observed
  if (length(ux$x) > length(troad))
  {
    print(dput(points))
    stop("Different length")
  }

  spline = cbind(ux$y, uy$y)
  spline = sf::st_sfc(sf::st_linestring(spline))

  return(spline)
}


R/find_seeds.R
#' Find seeds for the vectorization
#'
#' Drive along a contour line of the raster using \link{track_line} to find seeds that will be used
#' to vectorized the network
#'
#' @param map SpatRaster. Raster of probabilities/conductivities. Either a binary map or a probability
#' map the former being a special case of the later (see references)
#' @param contour an sfc_POLYGON that correspond to the contour of the raster. If the raster is
#' rectangular the default corresponds to the bounding box.
#' @param ... propagated to \link{track_line}
#'
#' @references
#' Jean-Romain Roussel, Jean-François Bourdon , Ilythia D. Morley , Nicholas C. Coops, Alexis Achim
#' (2022) Vectorial and topologically valid segmentation of forestry road networks from ALS data. In prep.\cr\cr
#' Jean-Romain Roussel, Jean-François Bourdon , Ilythia D. Morley , Nicholas C. Coops, Alexis Achim
#' (2022) Correction, update, and enhancement of vectorial forestry road maps using ALS data, a
#' pathfinder, and seven metrics. Journal of Applied Earth Observation and Geoinformation.
#'
#' @export
#' @examples
#' \dontrun{
#' library(terra)
#' library(sf)
#'
#' map <- system.file("extdata", "network.tif", package = "vecnet")
#' map <- rast(map)
#'
#' seeds = init_seeds(map, min_conductivity = 0.8)
#'
#' plot(map, col = viridis::inferno(25))
#' plot(seeds, add = TRUE, col = "green", lwd = 3)
#' }
init_seeds = function(map, contour = sf::st_as_sfc(sf::st_bbox(map)), ...)
{
  network = sf::st_buffer(contour, dist = -75)
  polygon = sf::st_difference(contour, network)
  network = sf::st_cast(network, "LINESTRING")
  bb = sf::st_cast(network, "POINT")
  seed = sf::st_linestring(sf::st_coordinates(bb[1:2]))
  l = sf::st_length(seed)
  r = 10/l
  seed = lwgeom::st_linesubstring(seed, 0, r)
  seed = sf::st_sfc(seed)
  sf::st_crs(seed) <- sf::st_crs(map)

  ans = track_line(seed, map, network, sightline = 50, find_seed_mode = polygon, ...)

  return(ans$seeds)
}



R/vectorize.R
#' Vectorize a raster representing a segmented roads network
#'
#' Vectorize a raster representing a segmented roads network. The raster can be either a binary
#' map or a non binary map (see examples). The algorithm is expected to be robust to gaps, and
#' false positive (see references).
#'
#' @param seeds  \code{sfc} a set of seeds to start driving the network (see references and examples).
#' Can be found with \link{init_seeds}.
#' @param map raster a probability raster.  Either a binary map or a probability map the
#' former being a special case of the later (see references).
#' @param network a \code{sf/sfc} with an already existing network such as the algorithm
#' can stop when vectorizing an already vectorized part of the network.
#' @param max_sinuosity numeric. False positive may be detected and discarded based on the fact that
#' they have likely an irrelevantly high sinuosity.
#' @param min_length numeric. False positive may be detected and discarded based on the fact that
#' they have likely an very short length.
#' @param smooth Boolean. By default the lines have one node per pixel. If true the lines are smoothed
#' and simplified. True by default.
#' @param verbose Boolean.
#' @param display Boolean. Display a map in pseudo real time of the advancement. For debugging purpose mainly.
#' @param ... propagated to \link{track_line}
#'
#' @references
#' Jean-Romain Roussel, Jean-François Bourdon , Ilythia D. Morley , Nicholas C. Coops, Alexis Achim
#' (2022) Vectorial and topologically valid segmentation of forestry road networks from ALS data. In prep.\cr\cr
#' Jean-Romain Roussel, Jean-François Bourdon , Ilythia D. Morley , Nicholas C. Coops, Alexis Achim
#' (2022) Correction, update, and enhancement of vectorial forestry road maps using ALS data, a
#' pathfinder, and seven metrics. Journal of Applied Earth Observation and Geoinformation.
#'
#' @examples
#' \dontrun{
#' library(terra)
#' library(sf)
#'
#' # =======================
#' # On a perfect binary map
#' # =======================
#'
#' # Generate a binary map from a vector for the example
#' network <- system.file("extdata", "network2.gpkg", package = "vecnet")
#' network <- buffer(vect(network), 1)
#' map <- rast(network, resolution = 2)
#' map <- rasterize(network, map)
#' map[is.na(map)] = 0.1
#' seeds <- system.file("extdata", "seeds2.gpkg", package = "vecnet")
#' seeds <- st_read(seeds) |> st_geometry()
#'
#' # Takes ~1 min to run
#' res <- vectorize_network(map, seeds, min_length = 150,  min_conductivity = 0.6, display = TRUE)
#'
#' # notice that some parts of the network are missing because they are not reachable
#' # with the seed used in this example.
#' col = sample(rainbow(length(res)))
#' plot(network, border = "gray")
#' plot(res, add = TRUE, col = col, lwd = 2)
#' starts = lwgeom::st_startpoint(res)
#' ends = lwgeom::st_endpoint(res)
#' plot(starts, add = TRUE, col = "black", pch = 19, cex = 0.5)
#' plot(ends, add = TRUE, col = "black", pch = 19, cex = 0.5)
#'
#' # =============================
#' # On an imperfect non binary map
#' # =============================
#'
#' map <- system.file("extdata", "network2.tif", package = "vecnet")
#' seeds <- system.file("extdata", "seeds2.gpkg", package = "vecnet")
#' map <- rast(map)
#' seeds <- st_read(seeds, quiet = TRUE) |> st_geometry()
#' plot(map, col = viridis::inferno(25), smooth = TRUE)
#' plot(seeds, add = T, col = "cyan", lwd = 4)
#'
#' res <- vectorize_network(map, seeds, min_conductivity = 0.6, display = TRUE)
#'
#' plot(map, col = viridis::inferno(25), smooth = TRUE, alpha = 0.5)
#' col = sample(rainbow(length(res)))
#' plot(res, add = TRUE, col = col, lwd = 2)
#' starts = lwgeom::st_startpoint(res)
#' ends = lwgeom::st_endpoint(res)
#' plot(starts, add = TRUE, col = "black", pch = 19, cex = 0.5)
#' plot(ends, add = TRUE, col = "black", pch = 19, cex = 0.5)
#' }
#' @export
vectorize_network = function(map, seeds, network = NULL, max_sinuosity = 1.8, min_length = 200, smooth = TRUE, verbose = FALSE, display = FALSE, ...)
{
  level <- 1
  col <- c("#FF0000", "#FF6600", "#FFCC00","#00FF00",
           "#00FF66", "#00FFCC", "#0066FF","#6600FF",
           "#CC00FF", "#FF0066")

  seeds = sf::st_geometry(seeds)

  if (display)
    terra::plot(map, col = viridis::inferno(25), smooth = TRUE)

  if (is.null(network))
  {
    network = seeds[0]
    netseeds = seeds[0]
  }


  while (!is.null(seeds))
  {
    ans <- vector("list", length(seeds))
    for (i in seq_along(seeds))
    {
      tryCatch({
        res <- track_line(seeds[i], map, network = network)#, ...)
      },
      error = function(e) {
        seed = seeds[i]
        f1 = tempfile(fileext = ".gpkg")
        f2 = tempfile(fileext = ".gpkg")
        sf::st_write(seed, f1)
        sf::st_write(network, f2)
        message(e)

        stop("Internal error. Reproduce with:
             seed = sf::st_read('", f1, "')
             netw = sf::st_read('", f2, "')
             track_line(seed, ", substitute(map), ", netw)" , call. = FALSE)
      })

      len <- as.numeric(sf::st_length(res$road))
      sin <- sinuosity(res$road)
      den <- res$dintersection

      if (len > min_length & sin <= max_sinuosity)
      {
        if (display) plot(sf::st_geometry(res$road), add = TRUE, lwd = 2, col = col[level])
        if(!is.null(res$seeds) && display) plot(sf::st_geometry(res$seeds), add = TRUE, lwd = 1, col = "lightblue")
        ans[[i]] <- res
        network = c(network, res$road)
        netseeds = c(netseeds, seeds[i])
      }
      else
      {
        if (len <= min_length && display)
          plot(sf::st_geometry(res$road), add = T, lwd = 2, col = "gray")

        if (sin > max_sinuosity && display)
          plot(sf::st_geometry(res$road), add = T, lwd = 2, col = "pink")
      }
    }

    ans <- Filter(Negate(is.null), ans)
    if (length(ans) == 0) break;

    seeds <- lapply(ans, function(x) x$seeds)
    seeds <- Filter(Negate(is.null), seeds)
    seeds <- do.call(c, seeds)

    level <- level +1
  }

  network <- sf::st_as_sf(network)
  #return(list(network, netseeds))

  #if (display)
  #{
  #  plot(network, add = F, col = sample(rainbow(nrow(network))), lwd = 2)
  #  starts = lwgeom::st_startpoint(network)
  #  ends = lwgeom::st_endpoint(network)
  #  plot(starts, add = TRUE, col = "black", pch = 19, cex = 0.5)
  #  plot(ends, add = TRUE, col = "black", pch = 19, cex = 0.5)
  #}

  network <- sf::st_geometry(network)
  network <- sfnetworks::as_sfnetwork(network)
  #plot(network)

  network <- tidygraph::convert(network, sfnetworks::to_spatial_subdivision)
  #plot(sf::st_geometry(network, "edges"), add = F, col = sample(rainbow(length(sf::st_geometry(network, "edges")))), lwd = 2)
  #plot(sf::st_geometry(network, "nodes"), add = TRUE, col = "black", pch = 19, cex = 0.5)

  network <- tidygraph::convert(network, sfnetworks::to_spatial_smooth)
  #plot(sf::st_geometry(network, "edges"), add = F, col = sample(rainbow(length(sf::st_geometry(network, "edges")))), lwd = 2)
  #plot(sf::st_geometry(network, "nodes"), add = TRUE, col = "black", pch = 19, cex = 0.5)

  network <- sf::st_geometry(network, "edges")
  #network <- sf::st_simplify(network, dTolerance = 2)

  if (smooth)
    network = smooth_network(network)

  return(network)
}

